### PR TITLE
[Functions] (feat): Request commitment hashes & metadata

### PR DESCRIPTION
**Pull request by justinkaseman**

### CODE REVIEW

File: contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsRequest.sol [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+pragma solidity ^0.8.19;
+
+interface IFunctionsRequest {
+  struct Commitment {
+    uint96 adminFee; // -----------┐
+    address coordinator; // -------┘
+    address client; // ------------┐
+    uint64 subscriptionId; //      |
+    uint32 callbackGasLimit; // ---┘
+    uint96 estimatedTotalCostJuels; // TODO pack the following
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsRequest.sol:11]: Are these comments still relevant or can they be removed?

File: contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsSubscriptions.sol [hunk 2]
```diff
@@ -77,9 +79,10 @@ interface IFunctionsSubscriptions {
 
   /**
    * @notice Time out all expired requests: unlocks funds and removes the ability for the request to be fulfilled
-   * @param requestIdsToTimeout - A list of request IDs to time out
+   * @param requestsToTimeoutByCommitment - A list of request commitments to time out
+   * @dev The commitment can be found on the "OracleRequest" event created when sending the request.
    */
-  function timeoutRequests(bytes32[] calldata requestIdsToTimeout) external;
+  function timeoutRequests(IFunctionsRequest.Commitment[] calldata requestsToTimeoutByCommitment) external;
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsSubscriptions.sol:85]: We will need to make the frontend team aware of this breaking change.

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsSubscriptions.sol [hunk 3]
```diff
@@ -508,35 +489,37 @@ abstract contract FunctionsSubscriptions is IFunctionsSubscriptions, ERC677Recei
   /**
    * @inheritdoc IFunctionsSubscriptions
    */
-  function timeoutRequests(bytes32[] calldata requestIdsToTimeout) external override {
+  function timeoutRequests(IFunctionsRequest.Commitment[] calldata requestsToTimeoutByCommitment) external override {
```
Comment (justinkaseman) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsSubscriptions.sol:492]: This is an awkward part about this change. Now to timeout requests the whole commitment has to be sent back and verified.

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol [hunk 4]
```diff
@@ -315,23 +303,25 @@ abstract contract FunctionsBilling is HasRouter, IFunctionsBilling {
...
-      billing.expectedGasPrice
+      s_config.gasOverheadBeforeCallback,
+      s_config.gasOverheadAfterCallback
     );
-    s_requestCommitments[requestId] = commitment;
 
-    emit BillingStart(requestId, commitment);
+    s_requestCommitments[requestId] = keccak256(abi.encode(commitment));
+
+    emit BillingStart(requestId, estimatedCost);
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:322]: Why do we emit only the estimated cost?  Would it be more helpful to emit the entire commitment?
Does it make sense to just omit this event entirely given all the data will also be contained in the OracleRequest event?

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol [hunk 5]
```diff
@@ -358,16 +348,12 @@ abstract contract FunctionsBilling is HasRouter, IFunctionsBilling {
...
-    if (commitment.don == address(0)) {
+    bytes memory err,
+    bytes memory onchainMetadata,
+    bytes memory /* offchainMetadata TODO: use in getDonFee() for dynamic billing */
+  ) internal returns (FulfillResult) {
+    IFunctionsRequest.Commitment memory commitment = _retrieveCommitmentData(onchainMetadata);
+    if (s_requestCommitments[requestId] != keccak256(abi.encode(commitment))) {
       return FulfillResult.INVALID_REQUEST_ID;
     }
     delete s_requestCommitments[requestId];
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:359]: I am still not convinced we should be deleting the request commitment until we confirm we have actually called the user's callback with the correct amount of gas. However, since that is outside of the scope of this PR, we can have the discussion elsewhere.

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol [hunk 6]
```diff
@@ -380,36 +366,46 @@ abstract contract FunctionsBilling is HasRouter, IFunctionsBilling {
...
-    emit BillingEnd(
-      requestId,
-      commitment.subscriptionId,
-      commitment.donFee,
-      gasOverheadJuels + callbackCostJuels,
-      gasOverheadJuels + callbackCostJuels + commitment.donFee + commitment.adminFee,
-      FulfillResult(result)
-    );
+    FulfillResult resultCode = FulfillResult(result);
+    if (resultCode == FulfillResult.USER_SUCCESS || resultCode == FulfillResult.USER_ERROR) {
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:391]: This is a significant change that should be considered carefully.
However, I think it makes sense to only emit a BillingEnd event if we have called the user's callback and I do not think this event is used in the core node or for obserability, so it is likely fine.

**Review by KuphJr — COMMENTED** [review]
Where is the corresponding core node code which provides the commitment back to the coordinator during report transmission?  Or is that expected in a subsequent PR?

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 4 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:322]: I took it out for that exact reason, it's already on OracleRequest.
I'm actually open to removing the billing events, we don't actually consume them anywhere.

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 5 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:359]: Not going to pile that into this PR, that is the concern of https://github.com/smartcontractkit/chainlink/pull/9984

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 6 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:391]: I'm just going to roll this back in favor of another PR

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 4 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:322]: Reverting this actually, to be handled on another PR

Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:322]: Yeah, if we can remove them I think we should. Less code = less problems, right?

File: contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsRequest.sol (hunk 1 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsRequest.sol:11]: Yes, still relevant. Still need to pack these more efficiently. I'll do it in another PR after running it by @RensR

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 6 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:391]: I'll leave this comment in case someone else thinks otherwise, but after considering this change for a bit, I think it is fine.

File: contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsSubscriptions.sol (hunk 2 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsSubscriptions.sol:85]: Do they even plan to handle timing out for Mainnet Preview?

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsSubscriptions.sol (hunk 3 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsSubscriptions.sol:492]: Yeah, not great but I don't see a way around it.

Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsSubscriptions.sol:492]: We are going to need to provide good support for this in the frontend.  I should probably be raised.

File: core/services/relay/evm/types/types.go [hunk 7]
```diff
@@ -53,6 +53,7 @@ type OracleRequest struct {
 	CallbackGasLimit    uint64
 	TxHash              common.Hash
 	CoordinatorContract common.Address
+	Commitment          []byte
```
Comment (bolekk) [inline core/services/relay/evm/types/types.go:56]: Let's call this OnchainMetadata please.

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol [hunk 8]
```diff
@@ -97,7 +86,7 @@ abstract contract FunctionsBilling is HasRouter, IFunctionsBilling {
   // ================================================================
   // |                         Cost Events                          |
   // ================================================================
-  event BillingStart(bytes32 indexed requestId, Commitment commitment);
+  event BillingStart(bytes32 indexed requestId, IFunctionsRequest.Commitment commitment);
```
Comment (pinebit) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:89]: Do we really need BillingStart/End events?
I don't see where it is useful, given that we have RequestStart/End carrying pretty much the same data.

File: contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsSubscriptions.sol (hunk 2 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsSubscriptions.sol:85]: If they don't plan on handling timeouts, then we are going to need to provide tooling to make it easy to do so.  This work needs to be added to the NPM package.

Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsSubscriptions.sol:85]: Ticket here: https://smartcontract-it.atlassian.net/browse/FUN-705

**Comment by KuphJr** [comment]
> Where is the corresponding core node code which provides the commitment back to the coordinator during report transmission? Or is that expected in a subsequent PR?

Followup, this is on line 164 of logpoller_wrapper.go

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 8 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:89]: I think that we can remove them, but I'll do so in a separate PR

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol [hunk 9]
```diff
@@ -388,18 +374,19 @@ abstract contract FunctionsBilling is HasRouter, IFunctionsBilling {
...
     // The Functions Router will perform the callback to the client contract
     IFunctionsRouter router = IFunctionsRouter(address(_getRouter()));
     (uint8 result, uint96 callbackCostJuels) = router.fulfill(
-      requestId,
       response,
       err,
       uint96(juelsPerGas),
-      costWithoutFulfillment,
-      msg.sender
+      gasOverheadJuels + commitment.donFee, // costWithoutFulfillment
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:387]: Can costWithoutFulfillment also include the admin fee?  Then we don't have to worry about accounting for the admin fee in the router's `fulfill` method.  I think this simplifies the code, but correct me if I am mistaken.

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsCoordinator.sol [hunk 10]
```diff
@@ -151,7 +153,7 @@ contract FunctionsCoordinator is OCR2Base, IFunctionsCoordinator, FunctionsBilli
     external
     override
     onlyRouter
-    returns (bytes32 requestId, uint96 estimatedCost, uint256 gasAfterPaymentCalculation, uint256 requestTimeoutSeconds)
+    returns (bytes32 requestId, uint96 estimatedCost, IFunctionsRequest.Commitment memory commitment)
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsCoordinator.sol:156]: `requestId` and `estimatedCost` are both within the commitment. Do we need to return them in a tuple or can we just return the commitment?

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol [hunk 11]
```diff
@@ -160,18 +155,25 @@ contract FunctionsRouter is RouterBase, IFunctionsRouter, FunctionsSubscriptions
...
-      coordinatorAddress,
-      msg.sender,
-      subscriptionId,
-      callbackGasLimit,
-      estimatedCost,
-      uint40(block.timestamp + requestTimeoutSeconds),
-      uint120(gasAfterPaymentCalculation)
+    s_requestCommitments[requestId] = keccak256(
+      abi.encode(
+        IFunctionsRequest.Commitment(
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol:163]: Why are we reconstructing the commitment?  Shouldn't we be using the commitment returned by `sendRequest`?

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsCoordinator.sol (hunk 10 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsCoordinator.sol:156]: Yeah good point. Tweaking it

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol (hunk 11 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol:163]: The router is using the information that it knows to be the truth, rather than trusting the what the coordinator sent

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsSubscriptions.sol [hunk 12]
```diff
@@ -508,14 +488,19 @@ abstract contract FunctionsSubscriptions is IFunctionsSubscriptions, ERC677Recei
...
+      // Check that request ID is valid
+      if (keccak256(abi.encode(request)) != s_requestCommitments[requestId]) {
+        revert InvalidCalldata();
+      }
 
       // Check that the message sender is the subscription owner
+      uint64 subscriptionId = request.subscriptionId;
       _isValidSubscription(subscriptionId);
       address owner = s_subscriptions[subscriptionId].owner;
       if (msg.sender != owner) {
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsSubscriptions.sol:506]: I thought we agreed anyone should be able to time out requests (or at the very least CLL should be able to): https://chainlink-core.slack.com/archives/C042Z3736JD/p1690914168586579

**Review by KuphJr — COMMENTED** [review]
Some questions & feedback

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 8 above)
Comment (bolekk) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:92]: If they're going then there's no need to extend them now.

File: contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsRequest.sol [hunk 13]
```diff
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+interface IFunctionsRequest {
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsRequest.sol:4]: Do we need a separate file because it's not used by both Router and Billing?

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol [hunk 14]
```diff
@@ -388,18 +370,19 @@ abstract contract FunctionsBilling is HasRouter, IFunctionsBilling {
     // (1e18 juels/link) * (wei/gas) / (wei/link) = juels per gas
     uint256 juelsPerGas = (1e18 * tx.gasprice) / uint256(weiPerUnitLink);
     // Gas overhead without callback
-    uint96 gasOverheadJuels = uint96(juelsPerGas * commitment.gasOverhead);
-    uint96 costWithoutFulfillment = gasOverheadJuels + commitment.donFee;
+    uint96 gasOverheadJuels = uint96(
+      juelsPerGas * (commitment.gasOverheadBeforeCallback + commitment.gasOverheadAfterCallback)
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:374]: Why is this changing? Some unrelated bugfix?

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol [hunk 15]
```diff
@@ -223,28 +223,26 @@ contract FunctionsRouter is RouterBase, IFunctionsRouter, FunctionsSubscriptions
...
     uint96 costWithoutFulfillment,
-    address transmitter
+    address transmitter,
+    IFunctionsRequest.Commitment memory commitment
   ) external override returns (uint8 resultCode, uint96 callbackGasCostJuels) {
     _whenNotPaused();
 
-    Commitment memory commitment = s_requestCommitments[requestId];
+    if (keccak256(abi.encode(commitment)) != s_requestCommitments[commitment.requestId]) {
+      resultCode = 2; // FulfillResult.INVALID_REQUEST_ID
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol:236]: I don't think that's a valid code here, can we introduce a new one?

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol [hunk 16]
```diff
@@ -223,28 +223,26 @@ contract FunctionsRouter is RouterBase, IFunctionsRouter, FunctionsSubscriptions
...
     }
 
-    if (commitment.client == address(0)) {
-      resultCode = 2; // FulfillResult.INVALID_REQUEST_ID
-      return (resultCode, callbackGasCostJuels);
-    }
-
     // Check that the transmitter has supplied enough gas for the callback to succeed
-    if (gasleft() < commitment.callbackGasLimit + commitment.gasAfterPaymentCalculation) {
+    if (gasleft() < commitment.callbackGasLimit + commitment.gasOverheadAfterCallback) {
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol:245]: ?

File: contracts/src/v0.8/functions/dev/1_0_0/RouterBase.sol [hunk 17]
```diff
@@ -49,8 +49,8 @@ abstract contract RouterBase is IRouterBase, Pausable, ITypeAndVersion, Confirme
     bytes to;
     uint256 timelockEndBlock;
   }
-
   mapping(bytes32 id => ConfigProposal) private s_proposedConfig;
+
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/1_0_0/RouterBase.sol:53]: unnecessary noise in the PR

**Review by bolekk — CHANGES_REQUESTED** [review]

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 14 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:374]: reducing one variable (costWithoutFulfillment) to help with callstack size
logic is the same

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol (hunk 16 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol:245]: It's the same, just using a consistent name from the Commitment interface

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 14 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:374]: previously, we just had a gasOverhead and now we have a gasOverheadAfterCallback and gasOverheadBeforeCallback. @justinkaseman correct me if I am wrong, but was this done just to make it easier to reason through the two places where we need to account for gas overhead?
ie: is the reason for splitting gasOverhad into 2 variables just for convenience?

File: contracts/src/v0.8/functions/dev/1_0_0/RouterBase.sol (hunk 17 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/RouterBase.sol:53]: sure, I can remove

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol (hunk 15 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol:236]: Yeah I think that makes more sense. Changing

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 8 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:92]: Sure. Usually it's preferred to have separation of concerns in PRs, but I'll remove them in this PR if you feel strongly about this line.

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 14 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:374]: Actually, I think I see the reason now.  Previously, we had 2 different Commitment structures
This was the one in FunctionsBilling
```
  struct Commitment {
    uint64 subscriptionId;
    address client;
    uint32 callbackGasLimit;
    address don;
    uint96 adminFee;
    uint96 estimatedTotalCostJuels;
    uint80 donFee;
    uint32 timestamp;
    uint40 gasOverhead;
    uint256 expectedGasPrice;
  }
```
And this was the on in FunctionsSubscriptions
```
  struct Commitment {
    uint96 adminFee;
    address coordinator;
    address client;
    uint64 subscriptionId;
    uint32 callbackGasLimit;
    uint96 estimatedCost;
    uint40 timeoutTimestamp;
    uint120 gasAfterPaymentCalculation;
  }
```

Now we have a unified Commitment structure in IFunctionsRequest
```
  struct Commitment {
    uint96 adminFee;
    address coordinator;
    address client;
    uint64 subscriptionId;
    uint32 callbackGasLimit;
    uint96 estimatedTotalCostJuels;
    uint40 timeoutTimestamp;
    bytes32 requestId;
    uint80 donFee;
    uint40 gasOverheadBeforeCallback; // old FunctionsBilling.Commitment.gasOverhead === new gasOverheadBeforeCallback + gasOverheadAfterCallback
    uint40 gasOverheadAfterCallback; // old FunctionsSubscriptions.Commitment.gasAfterPaymentCalculation === new gasOverhadAfterCallback
  }
}
```

Is this correct @justinkaseman

**Review by KuphJr — COMMENTED** [review]
I want to get answers to Bolek's comments, but I have nothing further to add. Getting close to approval.

File: contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsRequest.sol (hunk 13 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/1_0_0/interfaces/IFunctionsRequest.sol:4]: I agree with this.  I do prefer a single shared struct.

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol (hunk 16 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol:245]: Yep.  @bolekk see my comment: https://github.com/smartcontractkit/chainlink/pull/9898#discussion_r1281108875

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol (hunk 15 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol:236]: I think we can actually not even have requestId as a separate piece of the report now

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 14 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:374]: Yeah you have it right. It's just an effect from having a standard interface

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 8 above)
Comment (bolekk) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:92]: I didn't realize you have to add "IFunctionsRequest" prefix. Disregard and keep it, sorry.

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol (hunk 9 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:372]: Then we would be trusting the coordinator to tell the Router what to charge as the admin Fee, rather than taking it from the commitment data

**Review by KuphJr — APPROVED** [review]

**Review by KuphJr — DISMISSED** [review]

**Review by bolekk — APPROVED** [review]

**Review by KuphJr — APPROVED** [review]
